### PR TITLE
Word2vecSynonymFilter

### PR DESCRIPTION
**Pull request by dantuzi**
# Description

Please provide a short description of the changes you're making with this pull request.

# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [ ] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ ] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [ ] I have added tests for my changes.

### CODE REVIEW

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/Word2VecSynonymFilterFactory.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  @Override
+  public TokenStream create(TokenStream input) {
+    // if the synonymProvider is null, it means there's actually no synonyms... just return the original stream
+    return synonymProvider == null ? input : new Word2VecSynonymFilter(input, synonymProvider);
+  }
+
+  @Override
+  public void inform(ResourceLoader loader) throws IOException {
+    try(InputStream stream = loader.openResource(word2vecModel)) {
+      printJVMMemory("before reading file");
```
Comment (dantuzi) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/Word2VecSynonymFilterFactory.java]: remove print

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/Word2VecSynonymProvider.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+/**
+ * Implementation of a {@link SynonymProvider} using vector similarity technique
+ *
+ * @lucene.experimental
+ */
+public class Word2VecSynonymProvider implements SynonymProvider {
+
+    public static final VectorSimilarityFunction similarityFunction = VectorSimilarityFunction.COSINE;
+    public static final int MAX_CONN = 16;
```
Comment (dantuzi) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/Word2VecSynonymProvider.java]: org.apache.lucene.codecs.lucene90.Lucene90HnswVectorsFormat#DEFAULT_MAX_CONN

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/Word2VecSynonymProvider.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    public static final int BEAM_WIDTH = 10;
```
Comment (alessandrobenedetti) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/Word2VecSynonymProvider.java]: org.apache.lucene.codecs.lucene90.Lucene90HnswVectorsFormat#DEFAULT_BEAM_WIDTH

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/Word2VecSynonymProvider.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+import java.util.Map;
+
+/**
+ * Implementation of a {@link SynonymProvider} using vector similarity technique
+ *
+ * @lucene.experimental
+ */
+public class Word2VecSynonymProvider implements SynonymProvider {
+
+    public static final VectorSimilarityFunction similarityFunction = VectorSimilarityFunction.COSINE;
```
Comment (dantuzi) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/Word2VecSynonymProvider.java]: ref to dl4j

**Pull request closed without merging**